src/csharp/Microsoft.Spark/Sql/ForeachWriter.cs

```diff
@@ -0,0 +1,184 @@
+        /// <c>false</c>. However, <see cref="Close(Exception)"/> won't be called in the following
+        /// cases:
+        /// <list type="bullet">
+        /// <item>
+        /// <description>
+        /// CLR/JVM crashes without throwing a <see cref="Exception"/>.
+        /// </description>
+        /// </item>
+        /// <item>
+        /// <description>
+        /// <see cref="Open(long, long)"/> throws an <see cref="Exception"/>.
+        /// </description>
+        /// </item>
+        /// </list>
+        /// </summary>
+        /// <param name="errorOrNull">
+        /// The <see cref="Exception"/> thrown during processing or null if there was no error.
+        /// </param>
+        void Close(Exception errorOrNull);
+    }
+
+    /// <summary>
+    /// Wraps a <see cref="IForeachWriter"/> and calls the appropriate methods as decribed in
+    /// the lifecycle documentation for the interface.
+    /// </summary>
+    internal class ForeachWriterWrapper
+    {
+        private readonly IForeachWriter _foreachWriter;
+
+        internal ForeachWriterWrapper(IForeachWriter foreachWriter) =>
+            _foreachWriter = foreachWriter;
+
+        internal IEnumerable<object> Execute(int partitionId, IEnumerable<Row> rows)
+        {
+            if (!TaskContextHolder.Get().LocalProperties.TryGetValue(
+                "streaming.sql.batchId",
+                out string epochIdStr) || !long.TryParse(epochIdStr, out long epochId))
+            {
+                throw new Exception("Could not get batch id from TaskContext");
+            }
+
+            Exception error = null;
+            bool opened = _foreachWriter.Open(partitionId, epochId);
+
+            try
+            {
+                if (opened)
+                {
+                    foreach (Row row in rows)
+                    {
+                        _foreachWriter.Process(row);
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                error = e;
+            }
+            finally
+            {
+                _foreachWriter.Close(error);
+
+                if (error != null)
+                {
+                    throw error;
+                }
+            }
+
+            return Enumerable.Empty<object>();
+        }
+    }
+
+    /// <summary>
+    /// Wraps the given Func object, which represents a <see cref="ForeachWriterWrapper"/> UDF.
+    /// </summary>
+    [UdfWrapper]
+    internal class ForeachWriterWrapperUdfWrapper
+    {
+        private readonly Func<int, IEnumerable<Row>, IEnumerable<object>> _func;
+
+        internal ForeachWriterWrapperUdfWrapper(
+            Func<int, IEnumerable<Row>, IEnumerable<object>> func)
+        {
+            _func = func;
+        }
+
+        internal IEnumerable<object> Execute(int pid, IEnumerable<object> input)
+        {
+            return _func(
```

Comment:
can you add a comment what's happening here?

Reply:
related to `RowDeserializer`?

Reply:
Yes this is related to `RowDeserializer`.  The command execution goes through the `Spark.Utils.UdfUtils.PythonEvalType.NON_UDF` path, which expects a RDD.WorkerFunction.ExecuteDelegate Delegate.

Reply:
I meant to add it to the code :)

Reply:
I updated the `<summary>` portion of the class, instead of the `Execute` method.